File: FileReaderApp.c
//@! brief This program reads file content and print it to output console

#include <stdio.h>
#include <string.h>

#define FILE_PATH_POSITIVES        "/home/pi/CIATEQ/CompilersAndLibraries/testFilePositives.dat"
#define FILE_PATH_NEGATIVES        "/home/pi/CIATEQ/CompilersAndLibraries/testFileNegatives.dat"
#define FILE_PATH_FRACTIONAL       "/home/pi/CIATEQ/CompilersAndLibraries/testFileFractional.dat"
#define FILE_PATH_CHAR             "/home/pi/CIATEQ/CompilersAndLibraries/testFileChar.dat"
#define FILE_PATH_LONG_POSITIVES   "/home/pi/CIATEQ/CompilersAndLibraries/testFileLongPositives.dat"
#define FILE_PATH_LONG_NEGATIVES   "/home/pi/CIATEQ/CompilersAndLibraries/testFileLongNegatives.dat"
#define FILE_PATH_LONG_FRACTIONAL  "/home/pi/CIATEQ/CompilersAndLibraries/testFileLongFractional.dat"
#define FILE_PATH_LONG_CHAR        "/home/pi/CIATEQ/CompilersAndLibraries/testFileLongChar.dat"


int readmatrix(size_t rows, size_t cols, float (*a)[cols], const char* filename)
{

    FILE *pf;
    pf = fopen (filename, "r");
    if (pf == NULL)
        return 0;

    for(size_t i = 0; i < rows; ++i)
    {
        for(size_t j = 0; j < cols; ++j)
            fscanf(pf, "%f", a[i] + j);            
    }


    fclose (pf); 
    return 1; 
}

int main(void)
{
    int option = 0;
    float matrix[3][3];
    char lu8FileName[100]= {};
    
    printf("Please select an option: \r\n"); 
    printf("1: Test Positives \r\n");
    printf("2: Test Negatives \r\n");
    printf("3: Test Fractionals \r\n");
    printf("4: Test Chars \r\n");
    printf("5: Test Long Positives \r\n");
    printf("6: Test Long Negatives \r\n");
    printf("7: Test Long Fractionals \r\n");
    printf("8: Test Long Chars \r\n");
    
    scanf("%c", &option);
    
    switch(option)
    {
     case '1':
      strcpy((char*)lu8FileName, (const char*)FILE_PATH_POSITIVES);
     break;
     
     case '2':
      strcpy((char*)lu8FileName, (const char*)FILE_PATH_NEGATIVES);
     break;

     case '3':
      strcpy((char*)lu8FileName, (const char*)FILE_PATH_FRACTIONAL);
     break;
     
     case '4':
      strcpy((char*)lu8FileName, (const char*)FILE_PATH_CHAR);
     break;
     
     case '5':
      strcpy((char*)lu8FileName, (const char*)FILE_PATH_LONG_POSITIVES);
     break;
     
     case '6':
      strcpy((char*)lu8FileName, (const char*)FILE_PATH_LONG_NEGATIVES);
     break;

     case '7':
      strcpy((char*)lu8FileName, (const char*)FILE_PATH_LONG_FRACTIONAL);
     break;
     
     case '8':
      strcpy((char*)lu8FileName, (const char*)FILE_PATH_LONG_CHAR);
     break;
     
     default: 
      printf("Invalid option \r\n"); 
      return 0;
     break;
    
    }

    readmatrix(3, 3, matrix, (const char *)lu8FileName);
    
    printf("\r\n Input data: \r\n");
    for(size_t i = 0; i < 3; ++i)
    {
        for(size_t j = 0; j < 3; ++j)
            printf("%-3f ", matrix[i][j]);
        puts("");
    }

    printf("\r\n Output data: \r\n");
    for(size_t i = 0; i < 3; ++i)
    {
        for(size_t j = 0; j < 3; ++j)
            printf("%-3f ", (matrix[i][j]) * (matrix[i][j]) );
        puts("");
    }

    return 0;
}

File: FileReaderApp.py
import numpy as np
import time
import traceback
import sys


try:

 option = input("Please enter an option: \r\n"
 +"1: Test Positives\r\n"
 +"2: Test Negatives\r\n"
 +"3: Test Fractionals\r\n"
 +"4: Test Chars\r\n"
 +"5: Test Long Positives\r\n"
 +"6: Test Long Negatives\r\n"
 +"7: Test Long Fractionals\r\n"
 +"8: Test Long Chars\r\n"
 )
 
 if option == '1': 
  filePath = "C:\\Users\\EMMANUEL\\Academic\\CIATEQ\\Master degree\\TRIMESTER II\\Compilers\\Fortran programs\\testFilePositives.dat"
 elif option == '2': 
  filePath = "C:\\Users\\EMMANUEL\\Academic\\CIATEQ\\Master degree\\TRIMESTER II\\Compilers\\Fortran programs\\testFileNegatives.dat"
 elif option == '3': 
  filePath = "C:\\Users\\EMMANUEL\\Academic\\CIATEQ\\Master degree\\TRIMESTER II\\Compilers\\Fortran programs\\testFileFractional.dat"
 elif option == '4': 
  filePath = "C:\\Users\\EMMANUEL\\Academic\\CIATEQ\\Master degree\\TRIMESTER II\\Compilers\\Fortran programs\\testFileChar.dat"
 elif option == '5': 
  filePath = "C:\\Users\\EMMANUEL\\Academic\\CIATEQ\\Master degree\\TRIMESTER II\\Compilers\\Fortran programs\\testFileLongPositives.dat"
 elif option == '6': 
  filePath = "C:\\Users\\EMMANUEL\\Academic\\CIATEQ\\Master degree\\TRIMESTER II\\Compilers\\Fortran programs\\testFileLongNegatives.dat"
 elif option == '7': 
  filePath = "C:\\Users\\EMMANUEL\\Academic\\CIATEQ\\Master degree\\TRIMESTER II\\Compilers\\Fortran programs\\testFileLongFractional.dat"
 elif option == '8': 
  filePath = "C:\\Users\\EMMANUEL\\Academic\\CIATEQ\\Master degree\\TRIMESTER II\\Compilers\\Fortran programs\\testFileChar.dat"
 else:
  print("Invalid option") 
  sys.exit(0)
  
 #Get data from file
 np.fromfile(filePath, dtype=float)
 matrix = np.loadtxt(filePath, usecols=range(3))
 
 print ("Input data: \r\n")       
 print (matrix)
 
 print ("Output data: \r\n")        
 print (matrix*matrix)
 
 
 
 input("Press any key to continue")
 
except:
    traceback.print_exc()
    input("Press any key to continue")
